compiler/src/dotty/tools/dotc/transform/PatternMatcher.scala

```diff
@@ -802,7 +802,7 @@ object PatternMatcher {
           }
 
           expectedTp.dealias match {
-            case expectedTp: SingletonType =>
+            case SingletonType(expectedTp) =>
```

Comment:
This is wrong because the extractor for SingletonType will take only one branch of an AndType:
```scala
trait Foo {
  def hi: Int
}

object Test {
  def foo(x: Any) = x match {
    case x: ("1" & Foo) =>
      x.hi
    case _ =>
  }

  def main(args: Array[String]): Unit = {
    foo("1")
  }
}
```
This code used to run correctly, but now crashes with `java.lang.ClassCastException: java.lang.String cannot be cast to Foo`. In general I would rather not consider `AndType` to be singleton types, I think it's too easy to get incorrect semantics that way like what's happening here.

Reply:
I see. It looks easiest to revert and not let AndTypes be SingletonTypes, even though semantically it would be nice if they could be.